### PR TITLE
Provide meaningful config error message by tracking sub configurations

### LINKED ISSUE #1516: Use a bidirectional Map for Config to be able to retrieve full path to a property on error cases
## Expected Behavior
Full configuration path, e.g. rule set -> rule -> property, should be shown to help debug config errors.

## Current Behavior
When a wrong type is used for a property, only the property name is shown to the user.

### PR DESCRIPTION
**Pull request by arturbosch**
References:
- #1497
- #1498 

Now they look like:
`Value "[]" set for config parameter "RuleSet > Rule > active" is not of required type Int.`

Closes #1516

### CODE REVIEW

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt [hunk 1]
```diff
@@ -58,4 +57,26 @@ class ConfigSpec : Spek({
             YamlConfig.loadResource(javaClass.getResource("/oneitem.yml"))
         }
     }
+
+    describe("meaningful error messages") {
+
+        val config by memoized { yamlConfig("wrong-property-type.yml") }
+
+        it("prints whole config-key path for NumberFormatException") {
+            assertThatExceptionOfType(IllegalStateException::class.java).isThrownBy {
```
Comment (3flex) [inline detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt]: As above

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt [hunk 2]
```diff
@@ -58,4 +57,26 @@ class ConfigSpec : Spek({
...
+                config.subConfig("RuleSet")
+                    .subConfig("Rule")
+                    .valueOrDefault("threshold", 6)
+            }.withMessage("Value \"v5.7\" set for config parameter \"RuleSet > Rule > threshold\" is not of required type Int.")
+        }
+
+        it("prints whole config-key path for ClassCastException") {
+            assertThatExceptionOfType(IllegalStateException::class.java).isThrownBy {
```
Comment (3flex) [inline detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt]: As above

File: detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt [hunk 3]
```diff
@@ -41,10 +40,10 @@ class ConfigSpec : Spek({
         }
 
         it("tests wrong sub config conversion") {
-            assertThatExceptionOfType(ClassCastException::class.java).isThrownBy {
+            assertThatExceptionOfType(IllegalStateException::class.java).isThrownBy {
```
Comment (3flex) [inline detekt-api/src/test/kotlin/io/gitlab/arturbosch/detekt/api/ConfigSpec.kt]:
```suggestion
            assertThatIllegalStateException().isThrownBy {
```

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt [hunk 4]
```diff
@@ -44,43 +48,80 @@ interface Config {
...
+            Double::class,
+            String::class,
+            Short::class,
+            Char::class,
+            Long::class
+        )
     }
 }
 
+interface HierarchicalConfig : Config {
```
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt]: Memo: Needs a bit more documentation as this is public api.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt [hunk 5]
```diff
@@ -44,43 +48,80 @@ interface Config {
...
+                when {
+                    result is String -> tryParseBasedOnDefault(result, default)
+                    default::class in PRIMITIVES &&
+                        result::class != default::class -> throw ClassCastException()
                     else -> result
                 }
             } else {
                 default
             }
         } catch (e: ClassCastException) {
```
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt]: Memo: e is here also unused.

**Comment by 3flex** [comment]
Add "closes #1516" to description so issue is closed when this is merged?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt [hunk 6]
```diff
@@ -44,43 +48,80 @@ interface Config {
...
     }
 }
 
+interface HierarchicalConfig : Config {
+    /**
+     * Returns the parent config which encloses this config part.
+     */
+    val parent: Parent?
+
+    data class Parent(val config: Config, val key: String)
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Config.kt]: Maybe also add a comment here

**Pull request merged**